packages/flutter/lib/src/widgets/focus_scope.dart

```diff
@@ -1,145 +1,369 @@
-  static List<FocusScopeNode> ancestorsOf(BuildContext context) {
+  static FocusNode of(BuildContext context) {
     assert(context != null);
-    final List<FocusScopeNode> ancestors = <FocusScopeNode>[];
-    while (true) {
-      context = context.ancestorInheritedElementForWidgetOfExactType(_FocusScopeMarker);
-      if (context == null)
-        return ancestors;
-      final _FocusScopeMarker scope = context.widget;
-      ancestors.add(scope.node);
-      context.visitAncestorElements((Element parent) {
-        context = parent;
-        return false;
-      });
-    }
+    final _FocusMarker marker = context.inheritFromWidgetOfExactType(_FocusMarker);
+    return marker?.notifier ?? context.owner.focusManager.rootScope;
+  }
+
+  /// Returns true if the nearest enclosing [Focus] widget's node is focused.
+  ///
+  /// A convenience method to allow build methods to write:
+  /// `Focus.hasFocus(context)` to get whether or not the nearest [Focus] above
+  /// them in the widget hierarchy currently has the keyboard focus.
+  static bool hasFocus(BuildContext context) => Focus.of(context).hasFocus;
+
+  @override
+  void debugFillProperties(DiagnosticPropertiesBuilder properties) {
+    super.debugFillProperties(properties);
+    properties.add(StringProperty('debugLabel', debugLabel, defaultValue: null));
+    properties.add(FlagProperty('autofocus', value: autofocus, ifTrue: 'AUTOFOCUS', defaultValue: false));
+    properties.add(DiagnosticsProperty<FocusScopeNode>('node', focusNode, defaultValue: null));
   }
 
   @override
-  _FocusScopeState createState() => _FocusScopeState();
+  _FocusState createState() => _FocusState();
 }
 
-class _FocusScopeState extends State<FocusScope> {
+class _FocusState extends State<Focus> {
+  FocusNode _internalNode;
+  FocusNode get node => widget.focusNode ?? _internalNode;
+  bool _hasFocus;
   bool _didAutofocus = false;
 
-  @override
-  void didChangeDependencies() {
-    super.didChangeDependencies();
-    if (!_didAutofocus && widget.autofocus) {
-      FocusScope.of(context).setFirstFocus(widget.node);
-      _didAutofocus = true;
+  FocusNode _createNode() {
+    return FocusNode(
+      debugLabel: widget.debugLabel,
+    );
+  }
+
+  void _initNode() {
+    if (widget.focusNode == null) {
+      // Only create a new node if the widget doesn't have one.
+      _internalNode ??= _createNode();
     }
+    _hasFocus = node.hasFocus;
+    // Add listener even if the _internalNode existed before, since it should
+    // not be listening already if we're re-using a previous one.
+    node.addListener(_handleFocusChanged);
+    node.onKey = widget.onKey;
+  }
+
+  @override
+  void initState() {
+    super.initState();
+    _initNode();
   }
 
   @override
   void dispose() {
-    widget.node.detach();
+    // Regardless of the node owner, we need to remove it from the tree and stop
+    // listening to it.
+    node.removeListener(_handleFocusChanged);
+    node.onKey = null;
+    node.detach();
```

Comment:
This scares me. What if the node was passed to another widget, which has since attached to it?